### PR TITLE
feat(plugin-chart-echarts): support horizontal bar chart

### PR DESCRIPTION
**Pull request by stephenLYZ**
### SUMMARY

This PR supports horizontal bar chart for the new bar EChart. Here we switch these two bar orients (vertical and horizontal) by adding a `Bar Orient` control.

##### Vertical

image

##### Horizontal
image

##### Vertical
image

#### Horizontal
image

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://github.com/apache/superset/issues/15543#issuecomment-1100962800
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/Regular/Bar/controlPanel.tsx [hunk 1]
```diff
@@ -87,6 +89,28 @@ const config: ControlPanelConfig = {
     sections.advancedAnalyticsControls,
     sections.annotationsAndLayersControls,
     sections.forecastIntervalControls,
+    {
+      label: t('Chart Orient'),
```
Comment (rusackas) [inline superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/Regular/Bar/controlPanel.tsx:93]:
```suggestion
      label: t('Chart Orientation'),
```

**Comment by villebro** [comment]
I suspect this is one of the most anticipated PRs ever, love it! ❤️

**Comment by rusackas** [comment]
I think there are two little problems that need to be addressed in this PR. 

The small nitpicky one: 
I think having a whole _section_ for Bar orientation is a bit much... it leads to having "Chart orientation" and "Bar orientation" right next to each other, which is redundant. I'm wondering if it makes more sense to move the "Bar orientation" control into the "Chart options" section, rather than creating a new section.

And the trickier one:
When you transpose the chart between row and column orientation, the other controls may now confuse people. We refer to some of the controls as "x axis" and "y axis" which is no longer the case. I think we need one of two possible solutions here:
• Update the terminology, like "Metric axis" and "Dimension axis" to be more specific
• Updating the labeling of the controls so that the X and Y labels update on the fly depending on chart orientation, and the controls remain "truthful"

Thoughts @stephenLYZ or @villebro ?

**Comment by rusackas** [comment]
What my last comment didn't say is that this is AWESOME. Works great in `GENERIC_CHART_AXES` mode or the "normal" mode. I love that it even animates :D Pretty slick!

File: superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/transformProps.ts [hunk 2]
```diff
@@ -318,57 +322,68 @@ export default function transformProps(
...
-    },
+    xAxis,
+    yAxis,
     tooltip: {
       ...defaultTooltip,
       appendToBody: true,
       trigger: richTooltip ? 'axis' : 'item',
       formatter: (params: any) => {
+        const xIndex = isHorizontal ? 1 : 0;
+        const yIndex = isHorizontal ? 0 : 1;
```
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/transformProps.ts:374]: Slight simplification: 
```suggestion
        const [xIndex, yIndex] = isHorizontal ? [1, 0] : [0, 1];
```

**Review by villebro — COMMENTED** [review]
This is great work! A few small improvement proposals, other than that LGTM.

**Comment by stephenLYZ** [comment]
@rusackas Yeah. That makes sense! The reason that I prefer to make a whole section for Bar orientation is I'd like to have all the following x/y control swaps (include "Chart Title" section). Maybe we can implement it by dynamic control.

**Review by zhaoyongjie — APPROVED** [review]
The community has needed this feature for a long time, Thanks for addressing this!

**Comment by zhaoyongjie** [comment]
> And the trickier one: When you transpose the chart between row and column orientation, the other controls may now confuse people. We refer to some of the controls as "x axis" and "y axis" which is no longer the case.
> I think we need one of two possible solutions here:
> • Update the terminology, like "Metric axis" and "Dimension axis" to be more specific 
> • Updating the labeling of the controls so that the X and Y labels update on the fly depending on chart orientation, and the controls remain "truthful"

I totally agree, the label name of `x-axis` looks a bit misleading after it is rotated. The issue may not come from this PR, but from the current control panel design.
- the purpose of the `query panel` design is to describe the query fragment.
- the purpose of the `customize panel` design is to describe the properties of visualization.

The Bar chart transposition is a circumstance that a mixed query generator and viz definition.
1) In a vertical bar chart: the x-axis should be dimensions, it is represented by `x-axis` control and `dimension` control, y-axis should be metric(s).
2) In a horizontal bar chart: the x-axis should be metric(s). The `y-axis` should be dimensions, it is represented by x-axis control and dimension control.

Due to this confusing reason, lots of BI tools use `columns` and `rows` to represent `x-axis` and `y-axis` and provide many settings on each dimension/metric control.

image

**Comment by stephenLYZ** [comment]
@rusackas @zhaoyongjie That's a good input and currently it's based on the ability of the dynamic control to rotate - you can see that after setting the chart orientation, all x/y controls swap positions. Let's merge this one and improve this in the next PR.

**Comment by alldoami** [comment]
Looking through these comments... do I need to add a feature flag to my config to enable this?

**Comment by Vicinius** [comment]
Will this also work without time-series? Just with the regular bar chart? I just want to do a horizontal bar chart without time column... I'm asking here because I don't have acess to the superset version or upgrading it right now, with Superset 2.0 will I be able to do this?

**Comment by Chrisfufu** [comment]
> Will this also work without time-series? Just with the regular bar chart? I just want to do a horizontal bar chart without time column... I'm asking here because I don't have acess to the superset version or upgrading it right now, with Superset 2.0 will I be able to do this?

I also wondered the same. I am trying to do it without time-series now

**Comment by stephenLYZ** [comment]
@Vicinius Yeah. You can see this PR supports both Time-series and Caterical bar chart.

**Comment by redwards101** [comment]
Maybe I'm dense (entirely possible), but I can't get it to work. I have the GENERIC_CHART_AXES feature set in superset_config.py.  I'm using docker compose from near top of master, built the docker image to incorporate some other changes.  Can't get the v2 bar to function any differently. What am I missing?

**Comment by Vicinius** [comment]
@redwards101 see if this helps you https://github.com/apache/superset/issues/5059#issuecomment-1238517310

**Comment by redwards101** [comment]
> @redwards101 see if this helps you [#5059 (comment)](https://github.com/apache/superset/issues/5059#issuecomment-1238517310)

Thanks, but that's how I got here. There's clearly something I'm not doing right. Just don't know what it is yet. Looking for anyone who has it working in a docker environment.

**Comment by redwards101** [comment]
Ok, it's working now. I don't know why. I brought up the container again and suddenly I am now seeing the X-AXIS field that wasn't there before.  The only notable recent difference is I decided to just set the DEFAULT_FEATURE_FLAGS in config.py and rebuild the image.  I really can't see that as a reason, though.  Maybe something was cached?  Anyway, it's working great now.
Thanks for the feature.

**Comment by Vicinius** [comment]
@redwards101 were you able to sort the bars? that was a complaint on the comment that I pointed above

**Comment by redwards101** [comment]
> @redwards101 were you able to sort the bars? that was a complaint on the comment that I pointed above

No.  Can't control the axis sorting.

**Comment by didia** [comment]
> @Vicinius Yeah. You can see this PR supports both Time-series and Caterical bar chart.

@stephenLYZ Thank you so much for this feature. I can see the feature for the new V2 time series but the "normal" bar doesn't give option to switch orientation. Also, I'm not seeing a "normal" bar chart v2 in the latest superset version. Am I missing something ?

No orientation option in existing bar chart
Screen Shot 2022-10-06 at 01 42 34


No normal "V2" Bar chart
Screen Shot 2022-10-06 at 01 41 37

**Comment by Hugh-Guan** [comment]
> @Vicinius是的。您可以看到此 PR 同时支持时间序列和餐饮条形图。
Excuse me, I have upgraded superset to tag 2.0.0 version, and the horizontal control option has appeared in Time-series Bar Chart. But Bar Chart does not have this control, do I need to do any additional configuration?
caption：Time-series Bar Chart
image
caption：Bar Chart
image

**Comment by Hugh-Guan** [comment]
> > Will this also work without time-series? Just with the regular bar chart? I just want to do a horizontal bar chart without time column... I'm asking here because I don't have acess to the superset version or upgrading it right now, with Superset 2.0 will I be able to do this?
> 
> I also wondered the same. I am trying to do it without time-series now

@Vicinius I saw your question, and I thought the same thing. 
By the way, after superset was upgraded to TAG 2.0.0, did you implement horizontal function on your normal Bar Chart?

**Comment by Hugh-Guan** [comment]
> Hey @Hugh-Guan and @Vicinius , I'm not sure if you've been following the "generic x-axis" development work recently, but the "Time-series line/bar/scatter" etc charts from the ECharts chart plugin nowadays support using any x-axis type if you enable the `GENERIC_CHART_AXES` feature flag. With that flag enabled, I believe the ECharts bar chart already has feature parity with the old NVD3 based bar chart, and will soon hopefully deprecate the old one. If you are able to test master branch, please give it a spin and see if it fits your use cases.

@villebro First of all, I am very sorry for taking so long to reply. Then, thank you very much for your advice and help.
The GENERIC_CHART_AXES feature solved my problem perfectly.
image
image
Also, I noticed that Metrics are sorted in the natural order, not the order of choice. So far I've prefixed Metrics to show them in the order I want them to. I don't know if there is a function to configure it accordingly.
Thanks again for your help!

**Pull request merged**